numba/tests/test_array_analysis.py

```diff
@@ -0,0 +1,615 @@
+                # only test the equiv_set of the first block
+                func(analysis.equiv_sets[0])
+        if asserts == None:
+            self.assertTrue(self._has_no_assertcall(analysis.func_ir))
+        else:
+            for func in asserts:
+                func(analysis.func_ir, analysis.typemap)
+
+    def _has_assertcall(self, func_ir, typemap, args):
+        msg = "Sizes of {} do not match".format(', '.join(args))
+        for label, block in func_ir.blocks.items():
+            for expr in block.find_exprs(op='call'):
+                fn = func_ir.get_definition(expr.func.name)
+                if isinstance(fn, ir.Global) and fn.name == 'assert_equiv':
+                    typ = typemap[expr.args[0].name]
+                    if msg == typ.value:
+                        return True
+        return False
+
+    def _has_shapecall(self, func_ir, x):
+        for label, block in func_ir.blocks.items():
+            for expr in block.find_exprs(op='getattr'):
+                if expr.attr == 'shape':
+                    y = func_ir.get_definition(expr.value, lhs_only=True)
+                    z = func_ir.get_definition(x, lhs_only=True)
+                    y = y.name if isinstance(y, ir.Var) else y
+                    z = z.name if isinstance(z, ir.Var) else z
+                    if y == z:
+                        return True
+        return False
+
+    def _has_no_assertcall(self, func_ir):
+        for label, block in func_ir.blocks.items():
+            for expr in block.find_exprs(op='call'):
+                fn = func_ir.get_definition(expr.func.name)
+                if isinstance(fn, ir.Global) and fn.name == 'assert_equiv':
+                    return False
+        return True
+
+    def with_assert(self, *args):
+        return lambda func_ir, typemap: self.assertTrue(
+                        self._has_assertcall(func_ir, typemap, args))
+
+    def without_assert(self, *args):
+        return lambda func_ir, typemap: self.assertFalse(
+                        self._has_assertcall(func_ir, typemap, args))
+
+    def with_equiv(self, *args):
+        def check(equiv_set):
+            n = len(args)
+            for i in range(n-1):
+                if not equiv_set.is_equiv(args[i], args[n-1]):
+                    return False
+            return True
+        return lambda equiv_set: self.assertTrue(check(equiv_set))
+
+    def without_equiv(self, *args):
+        def check(equiv_set):
+            n = len(args)
+            for i in range(n-1):
+                if equiv_set.is_equiv(args[i], args[n-1]):
+                    return False
+            return True
+        return lambda equiv_set: self.assertTrue(check(equiv_set))
+
+    def with_shapecall(self, x):
+        return lambda func_ir, s: self.assertTrue(self._has_shapecall(func_ir, x))
+
+    def without_shapecall(self, x):
+        return lambda func_ir, s: self.assertFalse(self._has_shapecall(func_ir, x))
+
+    def test_base_cases(self):
```

Comment:
On a second thought, will it make more sense to move `test_9` to `test_parfors.py`? Since it is testing broadcast semantics, and it is the only test case in `test_array_analysis.py` that actually runs the program with `parallel=True`.